### PR TITLE
Expose ImproperUniform distribution

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2426 
Follows up on @fehiepsi's comment https://github.com/pyro-ppl/pyro/pull/2495#pullrequestreview-415691686

This exposes `ImproperUniform` as a public distribution. It is currently used only in `SplitReparam`. I plan to use it also for sampling auxiliary variables in `CompartmentalModel._sequential_model()` and `._vectorized_model()`. Those classes currently use `dist.Uniform(...).mask(False).expand(shape).to_event()`, but I would like to generalize the logic to arbitrary constraints, while adding the safety of a not-implemented `.sample()` method.

## Tested
- [x] refactoring is covered by existing `SplitReparam` tests
- [x] added some unit tests

### CODE REVIEW

File: pyro/distributions/improper_uniform.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from torch.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+from .util import broadcast_shape
+
+
+class ImproperUniform(TorchDistribution):
```
Comment (fehiepsi) [inline pyro/distributions/improper_uniform.py:11]: Just to make sure that you already consider it: should it be better to use the shorter name `Improper`?

File: pyro/distributions/improper_uniform.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        xyz = dist.Normal(0, 1).expand([3]).to_event(1).mask(False)
+
+    :param support: The support of the distribution.
+    :type support: ~torch.distributions.constraints.Constraint
+    :param torch.Size batch_shape: Defaults to empty.
+    :param torch.Size event_shape: Defaults to empty.
+    """
+    arg_constraints = {}
+
+    def __init__(self, support, batch_shape=torch.Size(), event_shape=torch.Size()):
```
Comment (fehiepsi) [inline pyro/distributions/improper_uniform.py:45]: I think `event_shape` is a requirement. Otherwise, we can set it to `None`, so that the default event_shape can be `(0,) * biject_to(support).event_dim`. This way, log_prob will have a more appropriate shape.

File: pyro/distributions/improper_uniform.py (hunk 1 above)
Comment (fritzo) [reply pyro/distributions/improper_uniform.py:11]: Hmm... I guess there are other improper distributions, e.g. `Unit`.

File: pyro/distributions/improper_uniform.py (hunk 2 above)
Comment (fritzo) [reply pyro/distributions/improper_uniform.py:45]: I think i'd rather make them both required than perform a computation on the constraint. Currently our constraints are pretty sloppy- we should be using `IndependentConstraint` more thoroughly and many of the computations would be wrong. This isn't a big issue yet because we don't use `biject_to(support).event_dim` for much.

Comment (fehiepsi) [reply pyro/distributions/improper_uniform.py]: Yeah, I agree that we should require event_shape. But for batch_shape, I guess it is better to let it have default shape as before (i.e. `__init__(self, support, event_shape, batch_shape=torch.Size())`) to makes the code less verbose.

Comment (fritzo) [reply pyro/distributions/improper_uniform.py]: I agree a default of `batch_shape=()` makes sense, but I find it disorienting for batch_shape to be specified on the right of event_shape: I always see things as `batch_shape, event_shape`. 🤷

Comment (fehiepsi) [reply pyro/distributions/improper_uniform.py]: Haha, sounds reasonable to me.

**Comment by fritzo** [comment]
Thanks for your careful review @fehiepsi!  I'm glad you care about interface elegance :smile:
